### ISSUE TITLE
On multi GPU Model picks up certain parameters from last run on Grid search

### ISSUE THREAD

**Issue by hadarishav** [PyTorchLightning/pytorch-lightning#2777]
I am doing a Grid Search over a set of hyperparameters. I do this by applying loops and re-initializing the model every time. However, the subsequent model seems to be picking up something from the previous runs because their performance changes as opposed to when I try only that particular setting of hyperparameters. Any idea what's happening here?

**Comment by Borda** [comment]
mind share an example?

**Comment by hadarishav** [comment]
@Borda Thanks for the quick reply. Here's an example - 
My code is as follows: 
`for i,lr in enumerate([2e-5]):
    for j,bert_dropout in enumerate([0.1]):
      for k,fc_dropout in enumerate([0.1,0.2]):

        ctr += 1
        config = {
        'PRE_TRAINED_MODEL_NAME': 'bert-base-cased',
        'batch_size': 16,
        'max_len': 200,
        'abuse_classes': 1,
        'bert_dropout': bert_dropout,
        'fc_dropout': fc_dropout,
        'device': torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu'),
        'num_epochs': 10,
        'lr': lr,
        'wd': 1e-4,
        'filename': 'dataset_10000.csv'

        }

        df = pd.read_csv(config['filename'])
        df = df.sample(frac=1).reset_index(drop=True)
        mini = min(df.Score)
        df_train, df_test = train_test_split(df, test_size=0.2, random_state=RANDOM_SEED)
        df_val, df_test = train_test_split(df_test, test_size=0.5, random_state=RANDOM_SEED)

        model = Abuse_lightning(mini, df_train,df_val,df_test, config)
        path = os.path.join(os.getcwd(), 'runs/lightning_logs/version_'+str(ctr)+'/checkpoints/')
        # print(path)
        checkpoint_callback = ModelCheckpoint(
          save_top_k=1,
          filepath = path,
          verbose=True,
          monitor='val_loss',
          mode='min')

        early_stop_callback = EarlyStopping(
         monitor='val_loss',
         min_delta=0.00,
         patience=3,
         verbose=False,
         mode='min'
      )
        trainer = pl.Trainer(gpus = 4, progress_bar_refresh_rate=0, max_epochs= config['num_epochs'], checkpoint_callback=checkpoint_callback, distributed_backend="ddp")
        trainer.fit(model)
  `
This gives me the following performance for the 2 sets of hyper parameters shown above: 
DATALOADER:0 TEST RESULTS
{'kendall': 0.029962238207975278,
 'loss': tensor(3.2553),
 'pearson': 0.043878178860853566,
 'spearman': 0.04405847290467181}

DATALOADER:0 TEST RESULTS
{'kendall': 0.5237109987894637,
 'loss': tensor(0.8089),
 'pearson': 0.7166115221028726,
 'spearman': 0.6918424244341342}

As you can see the 2nd set of hyperparameters is giving higher values. However, when I run only with those particular setting (i.e. not a grid search) performance is not the same, which should be the case. I get the following values:
DATALOADER:0 TEST RESULTS
{'kendall': 0.03853180875949515,
 'loss': tensor(0.9288),
 'pearson': 0.06812144454174138,
 'spearman': 0.05720029959393498}


So the next model is definitely picking something up from the last model. Hope this helps.

**Comment by hadarishav** [comment]
Turns out there was some problem with how I was handling the data. I will close this issue now. Thanks!

**Comment by hadarishav** [comment]
I am sorry for repeatedly opening and closing the issue. But I really wanted to be sure of the problem. So the problem was not really with how I was handling data. It is to do with lightning. If on multi gpu, I repeatedly run the same code on different files, the model picks something up from the previous runs in the lightning logs folder. If I delete that folder and do a subsequent run I get a certain performance, however when I let that folder be and do a subsequent run, the performance changes (becomes better). Here's the code:
```
if __name__ == "__main__":

  ctr = 0

  ctr += 1
  config = {
  'PRE_TRAINED_MODEL_NAME': 'bert-base-cased',
  'batch_size': 16,
  'max_len': 200,
  'abuse_classes': 1,
  'bert_dropout': 0.3,
  'fc_dropout': 0.3,
  'device': torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu'),
  'num_epochs': 10,
  'lr': 2e-5,
  'wd': 1e-4,
  'filename': ''

  }

  df_train = pd.read_csv("train.csv")
  df_val = pd.read_csv("val.csv")
  df_test = []

  model = Abuse_lightning(0, df_train,df_val,df_test, config)
  path = os.path.join(os.getcwd(), 'runs/lightning_logs/version_'+str(ctr)+'/checkpoints/')
  checkpoint_callback = ModelCheckpoint(
    save_top_k=1,
    filepath = path,
    verbose=True,
    monitor='val_loss',
    mode='min')

  early_stop_callback = EarlyStopping(
   monitor='val_loss',
   min_delta=0.00,
   patience=3,
   verbose=False,
   mode='min')
  
  trainer = pl.Trainer(gpus = 4, progress_bar_refresh_rate=0, max_epochs= config['num_epochs'], checkpoint_callback=checkpoint_callback, distributed_backend="ddp") #, distributed_backend="ddp"
  trainer.fit(model)
```
Now if I let the lightning_logs folder be, I get the following performance on 2 consecutive runs:
1st run - 
DATALOADER:0 TEST RESULTS
{'kendall': 0.060517550840717894,
 'loss': tensor(1.3496),
 'pearson': 0.09995590153171233,
 'spearman': 0.0893974875437095}

2nd run - 
DATALOADER:0 TEST RESULTS
{'kendall': 0.1193280845525937,
 'loss': tensor(0.9547),
 'pearson': 0.20193530230375267,
 'spearman': 0.17886979100553788}

If I delete the lightning_logs folder, I get the following performance on 2 consecutive runs:
1st run - 
 DATALOADER:0 TEST RESULTS
{'kendall': 0.060517550840717894,
 'loss': tensor(1.3496),
 'pearson': 0.09995590153171233,
 'spearman': 0.0893974875437095}

2nd run - 
DATALOADER:0 TEST RESULTS
{'kendall': 0.033294763979888835,
 'loss': tensor(1.0082),
 'pearson': 0.08154190029041658,
 'spearman': 0.04883269827105484}

As you can see, in the former case, the model performance is better in the 2nd run. 

Hope this helps!

**Comment by hadarishav** [comment]
I am sorry for repeatedly opening and closing the issue. But I really wanted to be sure of the problem. So the problem was not really with how I was handling data. It is to do with lightning. If on multi gpu, I repeatedly run the same code on different files, the model picks something up from the previous runs in the lightning logs folder. If I delete that folder and do a subsequent run I get a certain performance, however when I let that folder be and do a subsequent run, the performance changes (becomes better). Here's the code:
```
if __name__ == "__main__":

  ctr = 0

  ctr += 1
  config = {
  'PRE_TRAINED_MODEL_NAME': 'bert-base-cased',
  'batch_size': 16,
  'max_len': 200,
  'abuse_classes': 1,
  'bert_dropout': 0.3,
  'fc_dropout': 0.3,
  'device': torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu'),
  'num_epochs': 10,
  'lr': 2e-5,
  'wd': 1e-4,
  'filename': ''

  }

  df_train = pd.read_csv("train.csv")
  df_val = pd.read_csv("val.csv")
  df_test = []

  model = Abuse_lightning(0, df_train,df_val,df_test, config)
  path = os.path.join(os.getcwd(), 'runs/lightning_logs/version_'+str(ctr)+'/checkpoints/')
  checkpoint_callback = ModelCheckpoint(
    save_top_k=1,
    filepath = path,
    verbose=True,
    monitor='val_loss',
    mode='min')

  early_stop_callback = EarlyStopping(
   monitor='val_loss',
   min_delta=0.00,
   patience=3,
   verbose=False,
   mode='min')
  
  trainer = pl.Trainer(gpus = 4, progress_bar_refresh_rate=0, max_epochs= config['num_epochs'], checkpoint_callback=checkpoint_callback, distributed_backend="ddp") #, distributed_backend="ddp"
  trainer.fit(model)
```
Now if I let the lightning_logs folder be, I get the following performance on 2 consecutive runs:
1st run - 
DATALOADER:0 TEST RESULTS
{'kendall': 0.060517550840717894,
 'loss': tensor(1.3496),
 'pearson': 0.09995590153171233,
 'spearman': 0.0893974875437095}

2nd run - 
DATALOADER:0 TEST RESULTS
{'kendall': 0.1193280845525937,
 'loss': tensor(0.9547),
 'pearson': 0.20193530230375267,
 'spearman': 0.17886979100553788}

If I delete the lightning_logs folder, I get the following performance on 2 consecutive runs:
1st run - 
 DATALOADER:0 TEST RESULTS
{'kendall': 0.060517550840717894,
 'loss': tensor(1.3496),
 'pearson': 0.09995590153171233,
 'spearman': 0.0893974875437095}

2nd run - 
DATALOADER:0 TEST RESULTS
{'kendall': 0.033294763979888835,
 'loss': tensor(1.0082),
 'pearson': 0.08154190029041658,
 'spearman': 0.04883269827105484}

As you can see, in the former case, the model performance is better in the 2nd run. 

Hope this helps!

**Comment by awaelchli** [comment]
did you make your runs deterministic, e.g., `pytorch_lightning.seed_everything` ? Let's make sure this is the case.
In your two consecutive runs, what is the value of ctr ?

**Comment by hadarishav** [comment]
@awaelchli I did set the seed.
ctr was mainly for when gridsearch was done in a loop, so it would get updated each time.
However for debugging purpose, in the code I shared above I had removed the loops, and executed the code repeatedly myself. so ctr remained same for each run. 
In a new execution I would believe that data from past execution is not taken.

**Comment by awaelchli** [comment]
@hadarishav ok I think I see now, so ctr is the name of the version folder in your code. If that stays constant over multiple runs, then Lightning will resume from the last checkpoint because the version folder already contains a checkpoint.

**Comment by awaelchli** [comment]
I just checked, and on master, even if the checkpoint path already exists, it creates multiple versions with a prefix -v1, -v2, etc. 
If I restart, it always starts with the same val loss never continues from before, and it also saves new checkpoints.

**Comment by edenlightning** [comment]
closing this. please reopen if needed.
